File: 0146-lru-cache/0146-lru-cache.py
class Node:

    def __init__(self, key,val):        
        self.val = val
        self.key = key
        self.next = None
        self.prev = None

class LRUCache:
    hashMap = dict()
    capacity = 0
    def __init__(self, capacity: int):
        self.hashMap = dict()
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.capacity = capacity
        
    def removeNode(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev
    
    def addNode(self, node):
        node.next = self.head.next
        node.prev = self.head
        self.head.next = node
        node.next.prev = node


    def get(self, key: int) -> int:
        if key not in self.hashMap.keys():
            return -1
        
        currNode = self.hashMap.get(key)
        self.removeNode(currNode)
        self.addNode(currNode)
        return currNode.val
        

    def put(self, key: int, value: int) -> None:
        if key in self.hashMap.keys():
            newNode = Node(key, value)
            self.removeNode(self.hashMap.get(key))
            self.addNode(newNode)
            self.hashMap[key] = newNode
        else:
            if len(self.hashMap) == self.capacity:
                tailPrevNode = self.tail.prev
                self.removeNode(tailPrevNode)
                del self.hashMap[tailPrevNode.key]
            
            newNode = Node(key, value)
            self.addNode(newNode)
            self.hashMap[key] = newNode





# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 0146-lru-cache/0146-lru-cache.java
class Node{
    int key;
    int val;
    Node next;
    Node prev;

    public Node(int key, int val){
        this.key = key;
        this.val = val;
        // this.next = null;
        // this.prev = null;
    }
}

class LRUCache {
    int capacity;
    HashMap<Integer, Node> map;
    Node head;
    Node tail;

    public LRUCache(int capacity) {
        map = new HashMap<>();
        this.capacity = capacity;
        head = new Node(-1, -1);
        tail = new Node(-1, -1);
        head.next = tail;
        tail.prev = head;
        
    }
    
    private void removeNode(Node node){
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private void addNode(Node node){
        node.next = head.next;
        node.prev = head;
        head.next.prev = node;        
        head.next = node;

        // node.next = head.next;
        // node.prev = head;
        // head.next.prev = node;  // ✅ fix: move this BEFORE changing head.next
        // head.next = node;
    }

    public int get(int key) {
        if(!map.containsKey(key)){
            return -1;
        }
        
        Node currNode = map.get(key);
        // System.out.println(currNode.val);
        removeNode(currNode);
        addNode(currNode);
        return currNode.val;
    }
    
    public void put(int key, int value) {
        // System.out.println(value);
        if(map.containsKey(key)){
            Node node = map.get(key);
            node.val = value;
            removeNode(node);
            addNode(node);
        }else{
            if(map.size() == capacity){
                Node tailPrevNode = tail.prev;
                System.out.println(tailPrevNode.val);
                removeNode(tailPrevNode);
                map.remove(tailPrevNode.key);
            }
            System.out.println(value);
            System.out.println(map.size()+ "Size");
            Node newNode = new Node(key, value);
            addNode(newNode);
            map.put(key, newNode); 
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */